Reuse existing allocation id for primary shard allocation

ywelsch opened the pull request:
This solves an issue where shard was activated by master but the cluster crashes before the target node persisted the shard state.

Relates to #14739

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:
```diff
@@ -411,13 +411,17 @@ void moveToUnassigned(UnassignedInfo unassignedInfo) {
      * Initializes an unassigned shard on a node.
      */
-    void initialize(String nodeId, long expectedShardSize) {
+    void initialize(String nodeId, String existingAllocationId, long expectedShardSize) {
```
can we document that the existingAllocationId can be null?

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java:
```diff
@@ -698,6 +702,11 @@ public void initialize(String nodeId, long expectedShardSize) {
                 nodes.initialize(new ShardRouting(current), nodeId, expectedShardSize);
             }
 
+            public void initialize(String nodeId, String existingAllocationId, long expectedShardSize) {
```
copy over the java docs from the other method?

bleskes commented on core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java:
```diff
@@ -201,34 +200,34 @@ protected NodesResult buildAllocationIdBasedNodes(ShardRouting shard, boolean ma
+            logger.trace("{} candidates for allocation: {}", shard, nodeShardStates.stream().map(s -> s.getNode().getName()).collect(Collectors.joining(", ")));
         }
-        return new NodesResult(nodes, nodes.size());
+        return new NodeShardsResult(nodeShardStates, nodeShardStates.size());
```
I know it was already the case, but can we make this allocationFound parmater actually count the number of allocation found (i.e. all shards with an allocation id, matching or not) Later we can refactor this maybe and remove it completely but for now let's keep it consistent.

bleskes commented on core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java:
```diff
@@ -253,45 +252,44 @@ private boolean isEnoughVersionBasedAllocationsFound(ShardRouting shard, IndexMe
-        final Map<DiscoveryNode, Long> nodesWithVersion = new HashMap<>();
+    NodeShardsResult buildVersionBasedNodeShardsResult(ShardRouting shard, boolean matchAnyShard, Set<String> ignoreNodes,
+                                                       AsyncShardFetch.FetchResult<NodeGatewayStartedShards> shardState) {
+        final List<NodeGatewayStartedShards> nodeShardStateWithHighestVersion = new ArrayList<>();
```
this is getting confusing :) it is not nodes with highest version all the time. Can we call this allocationCandidates?

bleskes commented on core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java:
```diff
@@ -358,27 +347,28 @@ private boolean recoverOnAnyNode(IndexMetaData metaData) {
-    static class NodesResult {
-        public final List<DiscoveryNode> nodes;
+    static class NodeShardsResult {
+        public final List<NodeGatewayStartedShards> orderedNodeShardStates;
```
called this orderedAllocationCandidates?

bleskes commented on core/src/test/java/org/elasticsearch/cluster/routing/AllocationIdTests.java:
```diff
@@ -42,7 +42,7 @@ public void testShardToStarted() {
 
         logger.info("-- initialize the shard");
-        shard.initialize("node1", -1);
+        shard.initialize("node1", null, -1);
```
maybe just a add a variant which doesn't require this null?

bleskes commented:
Left some very minor comments...

ywelsch commented:
@bleskes Thanks for the review! I've updated the PR according to your suggestions and made one additional change: I removed the overloads of the `initialize` method that I earlier introduced in `ShardRouting`, `RoutingNodes` and `UnassignedShards` as there are not that many call sites for these methods. With each of these methods now being properly documented, I believe that adding `null` on these very few method calls to be cleaner (similar to `-1` that is present for `expectedShardSize` on most call sites).

bleskes commented:
LGTM. Thanks @ywelsch
